app/date-select/controller.test.js

```diff
@@ -0,0 +1,181 @@
+        expect(req.sessionModel.set).to.be.calledOnceWithExactly(
+          'referrer',
+          req.query.referrer
+        )
+      })
+
+      it('should redirect back to self', function () {
+        expect(res.redirect).to.be.calledOnceWithExactly(
+          req.form.options.fullPath
+        )
+      })
+    })
+
+    context('when called without a referrer', function () {
+      context('and a referrer has not been stored', function () {
+        beforeEach(function () {
+          controller.checkReferrer(req, res, next)
+        })
+
+        it('should not reset session model', function () {
+          expect(req.sessionModel.reset).to.not.be.called
+          expect(req.sessionModel.set).to.not.be.called
+        })
+
+        it('should get stored referrer', function () {
+          expect(req.sessionModel.get).to.be.calledOnceWithExactly('referrer')
+        })
+
+        it('should redirect back to home page', function () {
+          expect(res.redirect).to.be.calledOnceWithExactly('/')
+        })
+
+        it('should not call next', function () {
+          expect(next).to.not.be.called
+        })
+      })
+
+      context('and a referrer has been stored', function () {
+        beforeEach(function () {
+          req.sessionModel.get.returns('/referrer')
+          controller.checkReferrer(req, res, next)
+        })
+
+        it('should not reset session model', function () {
+          expect(req.sessionModel.reset).to.not.be.called
+          expect(req.sessionModel.set).to.not.be.called
+        })
+
+        it('should get stored referrer', function () {
+          expect(req.sessionModel.get).to.be.calledOnceWithExactly('referrer')
+        })
+
+        it('should call the super method', function () {
+          expect(next).to.be.calledOnceWithExactly()
+        })
+
+        it('should not redirect', function () {
+          expect(res.redirect).to.not.be.called
+        })
+      })
+    })
+  })
+
+  describe('#get', function () {
+    context('when called', function () {
+      beforeEach(function () {
+        req.sessionModel.get.returns('/referrer')
+        controller.get(req, res, next)
+      })
+
+      it('should get stored referrer', function () {
+        expect(req.sessionModel.get).to.be.calledOnceWithExactly('referrer')
+      })
+
+      it('should set the cancel and back links', function () {
+        expect(res.locals.cancelUrl).to.equal('/referrer')
+        expect(res.locals.backLink).to.equal('/referrer')
+      })
+
+      it('should call the super method', function () {
+        expect(FormWizardController.prototype.get).to.be.calledOnceWithExactly(
+          req,
+          res,
+          next
+        )
+      })
+    })
+  })
+
+  describe('#successHandler', function () {
+    context('when called', function () {
+      beforeEach(function () {
+        req.sessionModel.toJSON.returns({
+          referrer: '/bar/foo-1234/outgoing',
```

Comment:
Could these examples use actual dates, to make it clearer as to what is happening.
`2020-01-01` changing to `2020-12-01` is more meaningful that `foo-1234` changing to `stored-date`

Reply:
I wanted it to be clear that we are using a mocked value rather than the actual regex. Have hopefully now made it explicit enough.